src/wallet/wallet2.cpp

```diff
@@ -3153,6 +3153,60 @@ bool wallet2::verify_password(const std::string& keys_file_name, const epee::wip
   return r;
 }
 
+/*!
+ * \brief determine the key storage for the specified wallet file
+ * \param keys_file_name  Keys file to verify password for
+ * \param password        Password to verify
+ * \return                -1: incorrect password, 0 = default hw, 1 ledger hw
+ *
+ * for verification only - determines key storage hardware
+ *
+ */
+int wallet2::query_hardware(const std::string& keys_file_name, const epee::wipeable_string& password, uint64_t kdf_rounds)
+{
+  int key_on_device = 0;
+  rapidjson::Document json;
+  wallet2::keys_file_data keys_file_data;
+  std::string buf;
+  bool r = epee::file_io_utils::load_file_to_string(keys_file_name, buf);
+  THROW_WALLET_EXCEPTION_IF(!r, error::file_read_error, keys_file_name);
+
+  // Decrypt the contents
+  r = ::serialization::parse_binary(buf, keys_file_data);
+  THROW_WALLET_EXCEPTION_IF(!r, error::wallet_internal_error, "internal error: failed to deserialize \"" + keys_file_name + '\"');
+  crypto::chacha_key key;
+  crypto::generate_chacha_key(password.data(), password.size(), key, kdf_rounds);
+  std::string account_data;
+  account_data.resize(keys_file_data.account_data.size());
+  crypto::chacha20(keys_file_data.account_data.data(), keys_file_data.account_data.size(), key, keys_file_data.iv, &account_data[0]);
+  if (json.Parse(account_data.c_str()).HasParseError() || !json.IsObject())
+    crypto::chacha8(keys_file_data.account_data.data(), keys_file_data.account_data.size(), key, keys_file_data.iv, &account_data[0]);
+
+  // The contents should be JSON if the wallet follows the new format.
+  if (json.Parse(account_data.c_str()).HasParseError())
+  {
+    // old format before JSON wallet key file format
+  }
+  else
+  {
+    account_data = std::string(json["key_data"].GetString(), json["key_data"].GetString() +
+      json["key_data"].GetStringLength());
+
+    if (json.HasMember("key_on_device"))
+    {
+      GET_FIELD_FROM_JSON_RETURN_ON_ERROR(json, key_on_device, int, Int, false, false);
```

Comment:
You're assigning a bool to an int.

Reply:
AFAICT the JSON field is an int:

https://github.com/monero-project/monero/blob/b780cf4db1f9dfc49e7f16afc47892a5b40fe68a/src/wallet/wallet2.cpp#L2749-L2752

and

https://github.com/monero-project/monero/blob/b780cf4db1f9dfc49e7f16afc47892a5b40fe68a/src/wallet/wallet2.cpp#L2945-L2949

so the code is currently sticking an int into a bool. I would create an enum as you propose and use that in the JSON & change the type of ```m_key_on_device``` as well (and change all related code). Would that be ok?

Reply:
Ah, yes, it's rapidjson being weird with bools IIRC. It's fine now.

Reply:
As it is necessary to know what hardware the wallet is on, I will change m_key_on_device to an enum, esp. since m_key_on_device=true is hardwired to Ledger:

https://github.com/monero-project/monero/blob/b780cf4db1f9dfc49e7f16afc47892a5b40fe68a/src/wallet/wallet2.cpp#L3066-L3073

Reply:
It's still assigning a bool to an int fwiw. The last argument is a default value.

Reply:
oh!